lib/srv/db/cloud/users/users.go

```diff
@@ -0,0 +1,285 @@
+		// The extra jitter also helps offset small clock skews.
+		Duration: u.cfg.Interval * 7 / 6,
+	})
+	defer ticker.Stop()
+
+	for {
+		select {
+		case database := <-u.setupDatabaseChan:
+			u.setupDatabase(ctx, database)
+
+		case <-ticker.Next():
+			u.setupAllDatabases(ctx, getAllDatabases())
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// setupDatabase performs setup for a single database.
+func (u *Users) setupDatabase(ctx context.Context, database types.Database) {
+	// Database metadata is already refreshed once during database
+	// registration so no need to do it again.
+	u.setupDatabases(ctx, types.Databases{database}, false /*updateMeta*/)
+}
+
+// setupAllDatabases performs setup for all databases.
+func (u *Users) setupAllDatabases(ctx context.Context, allDatabases types.Databases) {
+	u.setupDatabases(ctx, allDatabases, true)
+
+	// Clean up.
+	u.lookup.removeUnusedDatabases(allDatabases)
+	activeUsers := u.lookup.usersByID()
+	for userID, user := range u.usersByID {
+		if _, found := activeUsers[userID]; !found {
+			delete(u.usersByID, userID)
+
+			if err := user.Teardown(ctx); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to tear down user %v.", user)
+			}
+		}
+	}
+}
+
+// setupDatabases performs setup for provided databases.
+func (u *Users) setupDatabases(ctx context.Context, databases types.Databases, updateMeta bool) {
+	for _, database := range databases {
+		fetcher, found := u.fetchersByType[database.GetType()]
+		if !found {
+			continue
+		}
+
+		// Refresh database metadata like ElastiCache user group IDs.
+		//
+		// Auto discovered databases can be skipped as the watcher service
+		// discovers changes periodically then triggers reconciler.
+		//
+		// If UpdateMeta fails, log an error and continue to next step with
+		// whatever meta database already has.
+		if updateMeta && database.Origin() != types.OriginCloud {
+			if err := u.cfg.UpdateMeta(ctx, database); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to update metadata for %q.", database)
+			}
+		}
+
+		// Fetch users.
+		fetchedUsers, err := fetcher.FetchDatabaseUsers(ctx, database)
+		if err != nil {
+			if trace.IsAccessDenied(err) { // Permission errors are expected.
+				u.cfg.Log.WithError(err).Debugf("No permissions to fetch users for %q.", database)
+			} else {
+				u.cfg.Log.WithError(err).Errorf("Failed to fetch users for database %v.", database)
+			}
+			continue
+		}
+
+		// Setup users.
+		var users []User
+		for _, fetchedUser := range fetchedUsers {
+			if user, err := u.setupUser(ctx, fetchedUser); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to setup user %v for database %v.", fetchedUser, database)
+			} else {
+				users = append(users, user)
+			}
+		}
+
+		// Rotate passwords.
+		for _, user := range users {
+			if err = user.RotatePassword(ctx); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to rotate password for user %v", user)
+			}
+		}
```

Comment:
rotation failures do not affect the `users` slice here, which is always used to update `database.GetManagedUsers`. So one failure will not affect the other.

Rotation failure should be rare if the setup succeeded since it should demand less. It is still possible though if network problems or maybe agents are racing in extreme cases (like in the same second). Though, rotation failure does not guarantee that GetPassword will fail. Most likely GetPasword is still fine just using the password successfully set in the previous round.

Thus, only logging an error here for rotation failure without any other action.